procrustes/kopt.py

```diff
@@ -20,148 +20,162 @@
+        raise ValueError(f"Argument k must be a integer greater than 2. Given k={k}")
+    if a.shape != b.shape:
+        raise ValueError(f"Argument b should have same shape as a. Given {b.shape} != {a.shape}")
+
+    # assign p1 & p2 to be an identity arrays, if not specified
+    n, m = a.shape
+    if p1 is None:
+        p1 = np.identity(n)
+    if p2 is None:
+        p2 = np.identity(m)
+
+    # compute 2-sided permutation error of the initial P & Q matrices
+    error = compute_error(a, b, p1, p2)
     # pylint: disable=too-many-nested-blocks
-    for comb_left in it.combinations(np.arange(num_row_left), r=kopt_k):
-        for comb_perm_left in it.permutations(comb_left, r=kopt_k):
-            if comb_perm_left != comb_left:
-                perm_kopt_left = deepcopy(perm_p)
-                # the right hand side permutation
-                for comb_right in it.combinations(np.arange(num_row_right), r=kopt_k):
-                    for comb_perm_right in it.permutations(comb_right, r=kopt_k):
-                        if comb_perm_right != comb_right:
-                            perm_kopt_right = deepcopy(perm_q)
-                            perm_kopt_right[comb_right, :] = perm_kopt_right[comb_perm_right, :]
-                            e_kopt_new_right = compute_error(array_n, array_m, perm_p.T,
-                                                             perm_kopt_right)
-                            if e_kopt_new_right < kopt_error:
-                                perm_q = perm_kopt_right
-                                kopt_error = e_kopt_new_right
-                                if kopt_error <= kopt_tol:
-                                    break
-
-                perm_kopt_left[comb_left, :] = perm_kopt_left[comb_perm_left, :]
-                e_kopt_new_left = compute_error(array_n, array_m, perm_kopt_left.T, perm_q)
-                if e_kopt_new_left < kopt_error:
-                    perm_p = perm_kopt_left
-                    kopt_error = e_kopt_new_left
-                    if kopt_error <= kopt_tol:
-                        break
-
-    return perm_p, perm_q, kopt_error
+
+    for perm1 in it.permutations(np.arange(n), r=k):
+        comb1 = sorted(perm1)
+        for perm2 in it.permutations(np.arange(m), r=k):
+            comb2 = sorted(perm2)
+            # permute rows of matrix P1
+            perm_p1 = np.copy(p1)
+            perm_p1[comb1, :] = perm_p1[perm1, :]
+            # permute rows of matrix P2
+            perm_p2 = np.copy(p2)
+            perm_p2[comb2, :] = perm_p2[perm2, :]
+            # compute error with new matrices & compare
+            perm_error = compute_error(b, a, perm_p1, perm_p2)
+            if perm_error < error:
+                p1, p2, error = perm_p1, perm_p2, perm_error
+                # check whether perfect permutation matrix is found
+                # TODO: smarter threshold based on norm of matrix
+                if error <= 1.0e-8:
+                    break
```

Comment:
Thank you for the updated codes! I only have a minor comment here.

We are not comparing or filtering the generated index here, which I think we can save some time (as I did in the old version). As far as I tested, if we are in the following situation,
```python
all = 0
equal_part = 0

p1 = np.identity(5)
p2 = np.identity(5)

for perm1 in it.permutations(np.arange(n), r=k):
    comb1 = tuple(sorted(perm1))
    for perm2 in it.permutations(np.arange(m), r=k):
        comb2 = tuple(sorted(perm2))
        if comb1 == perm1 and comb2 == perm2:
            equal_part += 1
        all += 1

print(all) # 3600
print(equal_part) # 100
```
where we will have 100 operations that are not swapping the columns/rows of left and right permutation matrix as the index are the same. That's saying we can save ~1/36 computation time if the comparison operation is not so expensive. 

I don't know if this necessary, but I was assuming it's going to decrease the computation time. Do you have any comments on this? Thanks!

@FarnazH @PaulWAyers

Reply:
Good points. No reason to do the extra work. Saving a factor of 1/k! (or 1/k!)**2 is always good, especially for k=2.

The bigger issue is re-writing the norm. I'd also like to be able to short-circuit the loop and do a "greedier" algorithm but re-loops through the permutations as soon as we are close. But .... I think those are new issues.

Reply:
Fixed in a57de40. Re-writing the norm will be done through issue #65.